lib/radius/dict.ex

```diff
@@ -1,327 +1,155 @@
-          DictEntry.__lookup__(__MODULE__, key)
-        end
-
-        def on_insert(val) do
-          val
-        end
-
-        def on_load(val) do
-          val
-        end
-
-        defoverridable on_insert: 1, on_load: 1
-      end
-    end
-
-    def __init__(mod) do
-      :ets.new(mod, [:set, :protected, :named_table])
-    end
-
-    def __insert__(mod, val) do
-      val = mod.on_insert(val)
-      keys = val |> mod.index_for
-
-      Enum.map(keys, fn x ->
-        if not :ets.insert_new(mod, {x, val}) do
-          # Logger.warn "ignored duplicat #{mod}: #{inspect x} at #{val.file}:#{val.line} \nObject: #{inspect val, pretty: true}"
-          :ok
-        end
-      end)
-    end
+        {acc_values, rest} =
+          case Map.pop(rest, :values) do
+            {[], rest} -> {acc_values, rest}
+            {values, rest} -> {values ++ acc_values, rest}
+          end
 
-    def __lookup__(mod, [key]) do
-      __lookup__(mod, key)
-    end
+        acc_vendors =
+          if vendor = Map.get(rest, :vendor) do
+            [vendor | acc_vendors]
+          else
+            acc_vendors
+          end
 
-    def __lookup__(mod, key) do
-      try do
-        :ets.lookup_element(mod, key, 2)
-      rescue
-        ArgumentError ->
-          raise EntryNotFoundError, type: mod, key: key
+        {acc_vendors, acc_attributes, acc_values}
       end
-    end
-  end
-
-  # module DictEntry
-
-  defmodule Vendor do
-    use DictEntry
-    defstruct id: nil, name: nil, format: {1, 1}, file: "(unknown)", line: 0
-
-    def index_for(val) do
-      [id: val.id, name: val.name]
-    end
-
-    def by_name(nil) do
-      %Vendor{}
-    end
-
-    def by_name(name) do
-      lookup!(name: name)
-    end
-
-    def by_id(id) do
-      lookup!(id: id)
-    end
-  end
-
-  # module Vendor
-
-  defmodule Attribute do
-    use DictEntry
-    defstruct id: nil, name: nil, type: nil, opts: [], vendor: nil, file: "(unknown)", line: 0
-
-    def on_insert(val) do
-      v = Vendor.by_name(val.vendor)
-      %{val | vendor: v}
-    end
-
-    def index_for(val) do
-      [id: {val.vendor.id, val.id}, name: val.name]
-    end
-
-    def by_name(name) do
-      lookup!(name: name)
-    end
-
-    def by_id(vendor \\ nil, id) do
-      lookup!(id: {vendor, id})
-    end
-  end
-
-  # module Attribute
-
-  defmodule Value do
-    use DictEntry
-    defstruct name: nil, attr: nil, value: nil, file: "(unknown)", line: 0
-
-    def on_insert(val) do
-      a = Attribute.lookup!(name: val.attr)
-      %{val | attr: a}
-    end
-
-    def index_for(val) do
-      [
-        value: {val.attr.vendor.id, val.attr.id, val.value},
-        name: {val.attr.vendor.name, val.attr.name, val.name}
-      ]
-    end
+    )
 
-    def by_name(vendor \\ nil, attr, name) do
-      lookup!(name: {vendor, attr, name})
-    end
+  for attribute <- attributes do
+    attr_fun_name = "attr_#{attribute.name}" |> String.replace("-", "_") |> String.to_atom()
```

Comment:
nitpick: `String.replace("-", "_")` should be extract as a `mangle_name` function

Reply:
Agree, in general I think this module needs to be cleaned up and a bit more DRY. However you would have to create the function before calling it at compile time.

If it's okay with you I'd rather merge and then polish in smaller chunks.